File: backend/app/api/v1/weather.py
"""
Weather Data API — Current and historical weather for cities
"""
from fastapi import APIRouter, Query
from typing import Optional
import random
import math
from app.data.cities_data import INDIAN_CITIES, get_city_by_id

router = APIRouter()


def generate_monthly_data(city: dict) -> list:
    """Generate realistic monthly weather data for a city."""
    months = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]
    base_temp = city["avg_temp"]
    base_humidity = city["humidity"]
    base_rainfall = city["rainfall_mm"] / 12

    monthly = []
    temp_pattern = [-4, -2, 2, 6, 9, 7, 3, 2, 1, -1, -3, -5]
    rain_pattern = [0.2, 0.3, 0.4, 0.5, 0.8, 2.5, 3.5, 3.2, 2.0, 1.2, 0.5, 0.2]
    humid_pattern = [-8, -6, -2, 2, 5, 12, 18, 16, 10, 2, -4, -8]

    for i, month in enumerate(months):
        temp = round(base_temp + temp_pattern[i] + random.uniform(-0.5, 0.5), 1)
        humidity = round(min(95, max(20, base_humidity + humid_pattern[i] + random.uniform(-2, 2))), 1)
        rainfall = round(max(0, base_rainfall * rain_pattern[i] + random.uniform(-5, 5)), 1)
        aqi = round(max(20, city["aqi"] + random.uniform(-30, 30) + (20 if i in [10, 11, 0] else -15 if i in [6, 7, 8] else 0)))

        monthly.append({
            "month": month,
            "month_num": i + 1,
            "avg_temp": temp,
            "max_temp": round(temp + 8 + random.uniform(0, 3), 1),
            "min_temp": round(temp - 6 + random.uniform(-2, 0), 1),
            "humidity": humidity,
            "rainfall_mm": rainfall,
            "aqi": aqi,
            "wind_speed_kmh": round(8 + random.uniform(-3, 8) + (5 if i in [5, 6] else 0), 1),
            "uv_index": round(min(12, max(3, 7 + math.sin(math.pi * (i - 2) / 6) * 4 + random.uniform(-0.5, 0.5))), 1),
        })
    return monthly


def generate_yearly_trend(city: dict) -> list:
    """Generate 10-year temperature trend."""
    trend = []
    base = city["avg_temp"] - 1.5
    for year in range(2015, 2026):
        warming = (year - 2015) * 0.15 + random.uniform(-0.2, 0.2)
        trend.append({
            "year": year,
            "avg_temp": round(base + warming, 1),
            "max_temp": round(city["max_temp"] - 2 + warming + random.uniform(-1, 1), 1),
            "aqi": round(max(30, city["aqi"] - 20 + (year - 2015) * 5 + random.uniform(-10, 10))),
            "green_cover_pct": round(max(3, city["green_cover_pct"] + 2 - (year - 2015) * 0.3 + random.uniform(-0.5, 0.5)), 1),
            "uhi_intensity": round(max(0.2, city["uhi_intensity"] - 0.5 + (year - 2015) * 0.08 + random.uniform(-0.1, 0.1)), 1),
        })
    return trend


@router.get("/current/{city_id}")
async def get_current_weather(city_id: str):
    """Get current weather conditions for a city."""
    city = get_city_by_id(city_id)
    if not city:
        return {"error": "City not found"}

    return {
        "city": city["name"],
        "state": city["state"],
        "temperature": round(city["avg_temp"] + random.uniform(-2, 5), 1),
        "feels_like": round(city["avg_temp"] + random.uniform(0, 8), 1),
        "humidity": city["humidity"],
        "wind_speed_kmh": round(8 + random.uniform(-3, 12), 1),
        "wind_direction": random.choice(["N", "NE", "E", "SE", "S", "SW", "W", "NW"]),
        "aqi": city["aqi"] + random.randint(-20, 20),
        "uv_index": round(random.uniform(5, 11), 1),
        "visibility_km": round(random.uniform(2, 10), 1),
        "pressure_hpa": round(1013 + random.uniform(-10, 10), 1),
        "lst": city["lst"],
        "conditions": random.choice(["Clear", "Partly Cloudy", "Hazy", "Sunny", "Hot"]),
    }


@router.get("/monthly/{city_id}")
async def get_monthly_weather(city_id: str):
    """Get monthly weather data for a city."""
    city = get_city_by_id(city_id)
    if not city:
        return {"error": "City not found"}
    random.seed(hash(city_id))
    return {"city": city["name"], "monthly_data": generate_monthly_data(city)}


@router.get("/yearly-trend/{city_id}")
async def get_yearly_trend(city_id: str):
    """Get yearly trend data (10 years)."""
    city = get_city_by_id(city_id)
    if not city:
        return {"error": "City not found"}
    random.seed(hash(city_id + "yearly"))
    return {"city": city["name"], "yearly_trend": generate_yearly_trend(city)}


@router.get("/heatwave-status")
async def heatwave_status():
    """Get current heatwave status across India."""
    random.seed(42)
    active_alerts = []
    for city in INDIAN_CITIES:
        if city["lst"] > 42:
            severity = "Extreme" if city["lst"] > 45 else "Severe" if city["lst"] > 43 else "Moderate"
            active_alerts.append({
                "city": city["name"],
                "state": city["state"],
                "lst": city["lst"],
                "severity": severity,
                "population_at_risk": round(city["population"] * random.uniform(0.2, 0.5)),
                "advisory": f"Extreme heat advisory for {city['name']}. Stay indoors between 11 AM - 4 PM.",
            })

    return {
        "total_alerts": len(active_alerts),
        "alerts": sorted(active_alerts, key=lambda x: x["lst"], reverse=True),
    }

File: backend/routes/weather.js
/**
 * Weather Routes — Current (LIVE), Monthly, Yearly, Heatwave
 */
const express = require('express');
const router = express.Router();
const { INDIAN_CITIES, getCityById } = require('../data/citiesData');
const { getCurrentWeather, convertAQI } = require('../services/liveWeather');

// Seeded random for consistent generated data
function seededRandom(seed) {
  let s = 0;
  for (let i = 0; i < seed.length; i++) s = ((s << 5) - s + seed.charCodeAt(i)) | 0;
  s = Math.abs(s) || 1;
  return () => { s = (s * 16807) % 2147483647; return (s & 2147483647) / 2147483647; };
}

function generateMonthlyData(city) {
  const months = ['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec'];
  const tempP = [-4,-2,2,6,9,7,3,2,1,-1,-3,-5];
  const rainP = [0.2,0.3,0.4,0.5,0.8,2.5,3.5,3.2,2.0,1.2,0.5,0.2];
  const humidP = [-8,-6,-2,2,5,12,18,16,10,2,-4,-8];
  const rng = seededRandom(city.id);
  const baseRain = city.rainfall_mm / 12;

  return months.map((month, i) => {
    const temp = Math.round((city.avg_temp + tempP[i] + (rng() - 0.5)) * 10) / 10;
    return {
      month, month_num: i + 1,
      avg_temp: temp,
      max_temp: Math.round((temp + 8 + rng() * 3) * 10) / 10,
      min_temp: Math.round((temp - 6 + (rng() - 1) * 2) * 10) / 10,
      humidity: Math.round(Math.min(95, Math.max(20, city.humidity + humidP[i] + (rng() - 0.5) * 4)) * 10) / 10,
      rainfall_mm: Math.round(Math.max(0, baseRain * rainP[i] + (rng() - 0.5) * 10) * 10) / 10,
      aqi: Math.round(Math.max(20, city.aqi + (rng() - 0.5) * 60 + ([10,11,0].includes(i) ? 20 : [6,7,8].includes(i) ? -15 : 0))),
      wind_speed_kmh: Math.round((8 + (rng() - 0.3) * 11 + ([5,6].includes(i) ? 5 : 0)) * 10) / 10,
      uv_index: Math.round(Math.min(12, Math.max(3, 7 + Math.sin(Math.PI * (i - 2) / 6) * 4 + (rng() - 0.5))) * 10) / 10,
    };
  });
}

function generateYearlyTrend(city) {
  const rng = seededRandom(city.id + 'yearly');
  const base = city.avg_temp - 1.5;
  return Array.from({ length: 11 }, (_, i) => {
    const year = 2015 + i;
    const warming = i * 0.15 + (rng() - 0.5) * 0.4;
    return {
      year, avg_temp: Math.round((base + warming) * 10) / 10,
      max_temp: Math.round((city.max_temp - 2 + warming + (rng() - 0.5) * 2) * 10) / 10,
      aqi: Math.round(Math.max(30, city.aqi - 20 + i * 5 + (rng() - 0.5) * 20)),
      green_cover_pct: Math.round(Math.max(3, city.green_cover_pct + 2 - i * 0.3 + (rng() - 0.5)) * 10) / 10,
      uhi_intensity: Math.round(Math.max(0.2, city.uhi_intensity - 0.5 + i * 0.08 + (rng() - 0.5) * 0.2) * 10) / 10,
    };
  });
}

// GET /current/:city_id — LIVE weather if API key set, else generated
router.get('/current/:city_id', async (req, res) => {
  const city = getCityById(req.params.city_id);
  if (!city) return res.status(404).json({ error: 'City not found' });

  // Try live data first
  const live = await getCurrentWeather(city.lat, city.lng);
  if (live) {
    return res.json({
      city: city.name, state: city.state,
      temperature: live.temperature, feels_like: live.feels_like,
      humidity: live.humidity, wind_speed_kmh: live.wind_speed_kmh,
      wind_direction: live.wind_direction, aqi: live.aqi_pm25 ? Math.round(live.aqi_pm25 * 2) : city.aqi,
      uv_index: 8, visibility_km: live.visibility_km,
      pressure_hpa: live.pressure_hpa, lst: city.lst,
      conditions: live.conditions, is_live: true,
    });
  }

  // Fallback to generated data
  const rng = seededRandom(city.id + Date.now().toString().slice(0, -4));
  res.json({
    city: city.name, state: city.state,
    temperature: Math.round((city.avg_temp + (rng() - 0.4) * 7) * 10) / 10,
    feels_like: Math.round((city.avg_temp + rng() * 8) * 10) / 10,
    humidity: city.humidity, wind_speed_kmh: Math.round((8 + (rng() - 0.2) * 15) * 10) / 10,
    wind_direction: ['N','NE','E','SE','S','SW','W','NW'][Math.floor(rng() * 8)],
    aqi: city.aqi + Math.floor((rng() - 0.5) * 40),
    uv_index: Math.round((5 + rng() * 6) * 10) / 10,
    visibility_km: Math.round((2 + rng() * 8) * 10) / 10,
    pressure_hpa: Math.round((1013 + (rng() - 0.5) * 20) * 10) / 10,
    lst: city.lst,
    conditions: ['Clear','Partly Cloudy','Hazy','Sunny','Hot'][Math.floor(rng() * 5)],
    is_live: false,
  });
});

router.get('/monthly/:city_id', (req, res) => {
  const city = getCityById(req.params.city_id);
  if (!city) return res.status(404).json({ error: 'City not found' });
  res.json({ city: city.name, monthly_data: generateMonthlyData(city) });
});

router.get('/yearly-trend/:city_id', (req, res) => {
  const city = getCityById(req.params.city_id);
  if (!city) return res.status(404).json({ error: 'City not found' });
  res.json({ city: city.name, yearly_trend: generateYearlyTrend(city) });
});

router.get('/heatwave-status', (req, res) => {
  const alerts = INDIAN_CITIES.filter(c => c.lst > 42).map(c => ({
    city: c.name, state: c.state, lst: c.lst,
    severity: c.lst > 45 ? 'Extreme' : c.lst > 43 ? 'Severe' : 'Moderate',
    population_at_risk: Math.round(c.population * (0.2 + Math.random() * 0.3)),
    advisory: `Extreme heat advisory for ${c.name}. Stay indoors between 11 AM - 4 PM.`,
  })).sort((a, b) => b.lst - a.lst);
  res.json({ total_alerts: alerts.length, alerts });
});

module.exports = router;
